### PR TITLE
Add processor frequency to profiler

### PR DESCRIPTION
**Pull request by AlexanderLemmens**
fix some linting warnings

### CODE REVIEW

File: triton-profiler/src/triton_profiler.rs [hunk 1]
```diff
@@ -75,9 +77,8 @@ impl TritonProfiler {
             self.stack.is_empty(),
             "Cannot generate report before stack is empty."
         );
-        assert_ne!(
-            Duration::ZERO,
-            self.total_time,
+        assert!(
+            self.total_time != Duration::ZERO,
```
Comment (jan-ferdinand) [inline triton-profiler/src/triton_profiler.rs:81]: This is probably not a desirable change.

File: triton-profiler/src/triton_profiler.rs [hunk 2]
```diff
@@ -176,7 +178,7 @@ impl TritonProfiler {
         });
 
         if std::env::var(GET_PROFILE_OUTPUT_AS_YOU_GO_ENV_VAR_NAME).is_ok() {
-            println!("start: {name}");
+            println!("stop: {name}; took {now:.2?}");
```
Comment (jan-ferdinand) [inline triton-profiler/src/triton_profiler.rs:181]:
```suggestion
            println!("start: {name}");
```

File: triton-profiler/src/triton_profiler.rs [hunk 3]
```diff
@@ -232,7 +234,7 @@ impl TritonProfiler {
         self.profile[index].time = duration;
 
         if std::env::var(GET_PROFILE_OUTPUT_AS_YOU_GO_ENV_VAR_NAME).is_ok() {
-            println!("stop:  {name} – took {duration:.2?}");
+            println!("stop: {name}");
```
Comment (jan-ferdinand) [inline triton-profiler/src/triton_profiler.rs:237]:
```suggestion
            println!("stop:  {name} – took {duration:.2?}");
```

File: triton-vm/Cargo.toml [hunk 4]
```diff
@@ -27,7 +27,7 @@ default-features = false
 
 [dependencies]
 twenty-first = { version = "0.7.2" }
-triton-profiler = { version = "0.7.2" }
+triton-profiler = { version = "0.7.2", path = "../triton-profiler" }
```
Comment (jan-ferdinand) [inline triton-vm/Cargo.toml:30]: We probably want users to use the crates.io package, not from the local path.

File: triton-vm/src/proof_item.rs [hunk 5]
```diff
@@ -119,8 +119,7 @@ pub enum ProofItem {
 impl MayBeUncast for ProofItem {
     fn uncast(&self) -> Vec<BFieldElement> {
         if let Self::Uncast(vector) = self {
-            let mut str = vec![];
-            str.push(BFieldElement::new(vector.len().try_into().unwrap()));
+            let mut str = vec![BFieldElement::new(vector.len().try_into().unwrap())];
```
Comment (jan-ferdinand) [inline triton-vm/src/proof_item.rs:122]: This is probably an improvement but doesn't really belong to the current scope. I have a slight (but not strong) preference to do it in a separate cleanup commit.

File: triton-profiler/src/triton_profiler.rs [hunk 6]
```diff
@@ -264,13 +266,16 @@ impl TritonProfiler {
...
-    pub fn finish(&mut self) {
+    pub fn finish(&mut self, cycle_count: Option<usize>) {
         assert!(!self.profile.is_empty(), "Nothing to finish.");
         assert!(
             self.stack.is_empty(),
             "Cannot finish before stack is empty."
         );
         self.total_time = self.timer.elapsed();
+        if let Some(cycle_count) = cycle_count {
+            self.clock_frequency = cycle_count / self.total_time.as_secs() as usize;
```
Comment (jan-ferdinand) [inline triton-profiler/src/triton_profiler.rs:277]: Be aware of division-by-zero errors here. This is also the reason the test suite failed.

**Review by jan-ferdinand — CHANGES_REQUESTED** [review]
Nice work! After having addressed the comments, I suggest to
1. squash the commits in this branch,
2. supply a descriptive commit message, and
3. rebase & fast-forward merge.

File: triton-profiler/src/triton_profiler.rs (hunk 1 above)
Comment (sshine) [inline triton-profiler/src/triton_profiler.rs:81]:
```suggestion
        assert_ne!(
            self.total_time,
            Duration::ZERO,
```

Comment (jan-ferdinand) [reply triton-profiler/src/triton_profiler.rs:81]: We usually do `assert_eq!(desired, actual)`. I'm not quite sure how that translates to `assert_ne!(…, …)` but usually go with `assert_ne!(undesired, actual)`.

**Comment by aszepieniec** [comment]
About merging finish and report: you might want to keep the option of separating them. For instance, if there is a benchmark in which not everything is being profiled. Then you want to tell the profiler, "stop tracking time now" but without spending cycles to prepare the report.

**Comment by jan-ferdinand** [comment]
> without spending cycles to prepare the report

Isn't it irrelevant when to spend the cycles if tracking has stopped?

**Comment by aszepieniec** [comment]
> > without spending cycles to prepare the report
> 
> Isn't it irrelevant when to spend the cycles if tracking has stopped?

Yes, we don't care (too much) about spending cycles after tracking has stopped. But one might want to profile a follow-up task or mother task with a separate profiler object, or even another benchmarker altogether. In this case, cycle-tracking is still on-going after the first `finish`.

It is a rather niche use case, but since the functionality is already there I just want to point out what we are making impossible without reverting changes.

File: triton-vm/Cargo.toml (hunk 4 above)
Comment (sshine) [reply triton-vm/Cargo.toml:30]: Correct. I'll change that now upon release; we needed to use the local one while developing.

**Pull request merged**